cpp/include/raft/cluster/kmeans_balanced.cuh

```diff
@@ -0,0 +1,366 @@
+/*
+ * Copyright (c) 2022, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <utility>
+
+#include <raft/cluster/detail/kmeans_balanced.cuh>
+#include <raft/core/mdarray.hpp>
+#include <raft/util/cuda_utils.cuh>
+
+namespace raft::cluster::kmeans_balanced {
+
+/**
+ * @brief Find clusters of balanced sizes with a hierarchical k-means algorithm.
+ *
+ * This variant of the k-means algorithm first clusters the dataset in mesoclusters, then clusters
+ * the subsets associated to each mesocluster into fine clusters, and finally runs a few k-means
+ * iterations over the whole dataset and with all the centroids to obtain the final clusters.
+ *
+ * Each k-means iteration applies expectation-maximization-balancing:
+ *  - Balancing: adjust centers for clusters that have a small number of entries. If the size of a
+ *    cluster is below a threshold, the center is moved towards a bigger cluster.
+ *  - Expectation: predict the labels (i.e find closest cluster centroid to each point)
+ *  - Maximization: calculate optimal centroids (i.e find the center of gravity of each cluster)
+ *
+ * The number of mesoclusters is chosen by rounding the square root of the number of clusters. E.g
+ * for 512 clusters, we would have 23 mesoclusters. The number of fine clusters per mesocluster is
+ * chosen proportionally to the number of points in each mesocluster.
+ *
+ * This variant of k-means uses random initialization and a fixed number of iterations, though
+ * iterations can be repeated if the balancing step moved the centroids.
+ *
+ * Additionally, this algorithm supports quantized datasets in arbitrary types but the core part of
+ * the algorithm will work with a floating-point type, hence a conversion function can be provided
```

Comment:
Just to make sure I'm up to date with the changes. Does that mean `MathT` must be floating point (`double` or `float`)? Or only `float`? (I vaguely remember a `predict_float_core` function here). Do we plan to support `half` here?
If we don't plan to support `half` in here, you can even use the `std::is_floating_point` static_assert here to be more pedantic :)

Reply:
`MathT = double` is supported and tested. Hence the renaming of `predict_float_core`. `half` could be supported in the future but there are a bunch of other prims you'll have to add `half` support for before you can do that.